pkg/controller/common.go

```diff
@@ -136,3 +172,349 @@ func cleanUpAzVolumeAttachments(ctx context.Context, client client.Client, azCli
+	if err != nil {
+		return nil, err
+	}
+
+	nodes := []string{}
+	for _, azVolumeAttachment := range azVolumeAttachments.Items {
+		nodes = append(nodes, azVolumeAttachment.Spec.NodeName)
+	}
+
+	return nodes, nil
+}
+
+func createLabelRequirements(label, value string) (*labels.Requirement, error) {
+	req, err := labels.NewRequirement(label, selection.Equals, []string{value})
+	if err != nil {
+		return nil, err
+	}
+	if req == nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("Unable to create Requirement to for label key : (%s) and label value: (%s)", label, value))
+	}
+	return req, nil
+}
+
+func createReplica(ctx context.Context, azr azReconciler, volumeID, node string) error {
+	diskName, err := azureutils.GetDiskNameFromAzureManagedDiskURI(volumeID)
+	if err != nil {
+		return err
+	}
+	volumeName := strings.ToLower(diskName)
+	replicaName := azureutils.GetAzVolumeAttachmentName(volumeName, node)
+
+	_, err = azr.getAzClient().DiskV1alpha1().AzVolumeAttachments(azr.getNamespace()).Create(ctx, &v1alpha1.AzVolumeAttachment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      replicaName,
+			Namespace: azr.getNamespace(),
+			Labels: map[string]string{
+				azureutils.NodeNameLabel:   node,
+				azureutils.VolumeNameLabel: volumeName,
+				azureutils.RoleLabel:       string(v1alpha1.ReplicaRole),
+			},
+		},
+		Spec: v1alpha1.AzVolumeAttachmentSpec{
+			NodeName:         node,
+			VolumeID:         volumeID,
+			UnderlyingVolume: volumeName,
+			RequestedRole:    v1alpha1.ReplicaRole,
+			VolumeContext:    map[string]string{},
+		},
+		Status: v1alpha1.AzVolumeAttachmentStatus{
+			State: v1alpha1.AttachmentPending,
+		},
+	}, metav1.CreateOptions{})
+
+	if err != nil {
+		klog.Infof("Replica AzVolumeAttachment (%s) has been successfully created.", replicaName)
+	}
+	return err
+}
+
+func getNamespacedName(namespace, name string) string {
+	return fmt.Sprintf("%s/%s", namespace, name)
+}
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func max(a, b int) int {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+func NewSharedMap() *SharedMap {
+	return &SharedMap{
+		mapping: map[string]interface{}{},
+		mutex:   sync.RWMutex{},
+	}
+}
+
+type SharedMap struct {
+	mapping map[string]interface{}
+	mutex   sync.RWMutex
+}
+
+func (m *SharedMap) get(key string) (interface{}, bool) {
+	m.mutex.RLock()
+	value, exists := m.mapping[key]
+	m.mutex.RUnlock()
+	return value, exists
+}
+
+func (m *SharedMap) getWithoutLock(key string) (interface{}, bool) {
+	value, exists := m.mapping[key]
+	return value, exists
+}
+
+func (m *SharedMap) getWithRetries(key string, maxRetry int, pollingRate time.Duration) (interface{}, bool) {
+	ok := false
+	var value interface{}
+	for i := 0; i < maxRetry && !ok; i++ {
+		value, ok = m.get(key)
+		time.Sleep(pollingRate)
+	}
+	return value, ok
+}
+
+func (m *SharedMap) add(key string, value interface{}) {
+	m.mutex.Lock()
+	m.mapping[key] = value
```

Comment:
This comment is irrelevant if you end up using `sync.Map` but this implementation would overwrite any existing entry with a new unlocked lock thereby defeating any previously added and locked instance. It should only add a new entry if one doesn't already exist.

Reply:
That was originally how I implemented the `add` function and had `update` function to do what the current `add` does but there was no real use case for the original `add` so I got rid of it :(

Reply:
Ah, I was examining this implementation in context of storing `Mutex` values (specifically for `mutexMap` on in `AzVolume` an replica controllers). This pattern is unsafe when used with `Mutex` values and easy to misuse.

Reply:
Ah, I see what you mean. :) Thanks for the clarification.